document-store/src/integrationTest/java/org/hypertrace/core/documentstore/DocStoreTest.java

```diff
@@ -1055,6 +1061,188 @@ public void testSearch(String dataStoreName) throws IOException {
     }
   }
 
+  @ParameterizedTest
+  @MethodSource("databaseContextProvider")
+  public void testCreateWithMultipleThreads(String dataStoreName) throws Exception {
+    Datastore datastore = datastoreMap.get(dataStoreName);
+    Collection collection = datastore.getCollection(COLLECTION_NAME);
+    SingleValueKey documentKey = new SingleValueKey("default", "testKey1");
+
+    Map<String, List<CreateUpdateTestThread>> resultMap =
+        executeCreateUpdateThreads(collection, Operation.CREATE, 5, documentKey);
+
+    Assertions.assertEquals(1, resultMap.get(SUCCESS).size());
+    Assertions.assertEquals(4, resultMap.get(FAILURE).size());
+
+    // check the inserted document and thread result matches
+    Query query = new Query();
+    query.setFilter(Filter.eq("_id", documentKey.toString()));
+    Iterator<Document> results = collection.search(query);
+    List<Document> documents = new ArrayList<>();
+    while (results.hasNext()) {
+      documents.add(results.next());
+    }
+    Assertions.assertTrue(documents.size() == 1);
+    Map<String, Object> doc = OBJECT_MAPPER.readValue(documents.get(0).toJson(), Map.class);
+    Assertions.assertEquals(resultMap.get(SUCCESS).get(0).getTestValue(), (int) doc.get("size"));
+  }
+
+  @ParameterizedTest
+  @MethodSource("databaseContextProvider")
+  public void testUpdateWithMultipleThreads(String dataStoreName) throws Exception {
```

Comment:
Should we include a test for concurrently update where the first one gets updated the rest fail? Maybe it is already in but I can't not easy identify such test from reading them.

Reply:
Yes. this one is doing the same.
- create one document
- `5` threads trying to update with the condition, and only one success, and for same I am using test class `CreateUpdateTestThread`